src/base/misc/fatfs.c

```diff
@@ -424,26 +454,44 @@ static void set_geometry(fatfs_t *f, unsigned char *b)
   bpb->sectors_per_track = f->secs_track;
   bpb->num_heads = f->heads;
 
-  if (bpb->v340_400_signature == BPB_SIG_V340 ||
-      bpb->v340_400_signature == BPB_SIG_V400) {
-    bpb->v331_400_hidden_sectors = f->hidden_secs;;
-    bpb->v331_400_num_sectors_large = (f->total_secs < 65536L) ? 0 : f->total_secs;
-    bpb->v340_400_drive_number = f->drive_num;
-
-  } else if (f->sys_type == OLDDRD_D) {    // DR-DOS 3.40 / 3.41 has v3.20 BPB
-    bpb->v300_320_hidden_sectors = f->hidden_secs;
-    bpb->v320_num_sectors_large = (f->total_secs < 65536L) ? 0 : f->total_secs;
-    // drive_number not passed in boot block as string data in 0x1fd position
-
-  } else {                        // Could be any of v3.00, v3.20 or v3.31 BPB
-    if (bpb->num_sectors_small > 0) { // FAT12 or FAT16
-      // Not compatible with v3.31, but replicates earlier code
+  /* set the geometry with BPB version dependent fields */
+  if (bpb->num_sectors_small > 0) { // FAT12 or FAT16
+    if (version < 331) {
+      // Since small disks and early versions of BPB can be indistinguishable,
+      // we can't be as specific as we would like. But we don't need to be as
+      // we need to write the only first uint16 i.e v300_320_hidden_sectors
+      // and the following will be zeros as before.
       bpb->v300_320_hidden_sectors = f->hidden_secs;
-    } else {                          // FAT16B
-      // Likely v3.31, not compatible with also possible v3.20
-      bpb->v331_400_hidden_sectors = f->hidden_secs;;
-      bpb->v331_400_num_sectors_large = (f->total_secs < 65536L) ? 0 : f->total_secs;
+    } else {
+      bpb->v331_400_hidden_sectors = f->hidden_secs;
+      bpb->v331_400_num_sectors_large = 0;
+    }
+  } else {                          // FAT16B
+    if (version == 300) {
+      // Misidentification of the boot block might have occured. This could
```

Comment:
Uhhh, I don't want to even look into all this crap! :)
Please just move your crazy things elsewhere, like
in convert_boot_blk(), and please leave the generic
set_geometry() path untouched (or simplified, it doesn't
need to be fancy).